test/aleph/http_test.clj

```diff
@@ -1308,3 +1312,107 @@
+        (is (instance? ExceptionInfo result))
+        (is (= "Some desired HTTP versions are not part of ALPN config." (ex-message result)))))
+    (testing "with different preference order in ALPN config"
+      (let [result (test-start-server-options
+                    {:http-versions [:http2 :http1]
+                     :ssl-context (netty/coerce-ssl-server-context
+                                   (assoc test-ssl/server-ssl-context-opts
+                                          :application-protocol-config
+                                          (netty/application-protocol-config [:http1 :http2])))})]
+        (is (instance? ExceptionInfo result))
+        (is (= "Desired HTTP version preference order differs from ALPN config." (ex-message result)))))
+    (testing "with extra HTTP versions in the ALPN config"
+      (let [result (test-start-server-options
+                    {:http-versions [:http1]
+                     :ssl-context (netty/coerce-ssl-server-context
+                                   (assoc test-ssl/server-ssl-context-opts
+                                          :application-protocol-config
+                                          (netty/application-protocol-config [:http1 :http2])))})]
+        (is (instance? ExceptionInfo result))
+        (is (= "ALPN config contains more HTTP versions than desired." (ex-message result)))))
+    (testing "with matching ALPN config"
+      (let [result (test-start-server-options
+                    {:http-versions [:http2 :http1]
+                     :ssl-context (netty/coerce-ssl-server-context
+                                   (assoc test-ssl/server-ssl-context-opts
+                                          :application-protocol-config
+                                          (netty/application-protocol-config [:http2 :http1])))})]
+        (is (= :started result))))
+    (testing "with no ALPN config"
+      (let [result (test-start-server-options
+                    {:http-versions [:http2 :http1]
+                     :ssl-context test-ssl/server-ssl-context})]
+        (is (instance? ExceptionInfo result))
+        (is (= "Some desired HTTP versions are not part of ALPN config." (ex-message result))))))
+  (testing "HTTP/2 without ssl-context"
+    (let [result (test-start-server-options
+                  {:http-versions [:http2]})]
+      ;; TODO: Shouldn't this fail?
+      (is (= :started result))))
+  (testing "HTTP/2 without ssl-context but with h2c"
+    (let [result (test-start-server-options
+                  {:use-h2c? true
+                   :http-versions [:http2]})]
+      (is (= :started result))))
+  (testing "HTTP/2 with ssl-context *and* with h2c (which has no effect then)"
+    (let [result (test-start-server-options
+                  {:use-h2c? true
+                   :http-versions [:http2]
+                   :ssl-context test-ssl/server-ssl-context-opts})]
+      (is (= :started result)))))
```

Comment:
Should also throw an error, because of port incompatibilities, if nothing else.

Reply:
Done! And while at it, I also made it an error to enable h2c without enabling H2 (in a separate commit, so easy to drop if you disgree with that one).